src/Symfony/Bundle/FrameworkBundle/Kernel/MicroKernelTrait.php

```diff
@@ -149,28 +156,42 @@ public function loadRoutes(LoaderInterface $loader)
         $collection = new RouteCollection();
 
         try {
-            $this->configureRoutes(new RoutingConfigurator($collection, $kernelLoader, $file, $file));
+            $configureRoutes = new \ReflectionMethod($this, 'configureRoutes');
+        } catch (\ReflectionException $e) {
+            throw new \LogicException(sprintf('"%s" uses "%s", but does not implement the required method "protected function configureRoutes(RoutingConfigurator $routes): void".', get_debug_type($this), MicroKernelTrait::class), 0, $e);
+        }
 
-            foreach ($collection as $route) {
-                $controller = $route->getDefault('_controller');
+        if (($configuratorClass = $this->getParameterClassName($configureRoutes, 0)) && !is_a(RoutingConfigurator::class, $configuratorClass, true)) {
+            trigger_deprecation('symfony/framework-bundle', '5.1', 'Using type "%s" for argument 1 of method "%s:configureRoutes()" is deprecated, use "%s" instead.', RouteCollectionBuilder::class, self::class, RoutingConfigurator::class);
 
-                if (\is_array($controller) && [0, 1] === array_keys($controller) && $this === $controller[0]) {
-                    $route->setDefault('_controller', ['kernel', $controller[1]]);
-                }
-            }
+            $routes = new RouteCollectionBuilder($loader);
+            $this->configureRoutes($routes);
 
-            return $collection;
-        } catch (\TypeError $e) {
-            if (0 !== strpos($e->getMessage(), sprintf('Argument 1 passed to %s::configureRoutes() must be an instance of %s,', static::class, RouteCollectionBuilder::class))) {
-                throw $e;
+            return $routes->build();
+        }
+
+        $this->configureRoutes(new RoutingConfigurator($collection, $kernelLoader, $file, $file));
+
+        foreach ($collection as $route) {
+            $controller = $route->getDefault('_controller');
+
+            if (\is_array($controller) && [0, 1] === array_keys($controller) && $this === $controller[0]) {
+                $route->setDefault('_controller', ['kernel', $controller[1]]);
             }
         }
 
-        trigger_deprecation('symfony/framework-bundle', '5.1', 'Using type "%s" for argument 1 of method "%s:configureRoutes()" is deprecated, use "%s" instead.', RouteCollectionBuilder::class, self::class, RoutingConfigurator::class);
+        return $collection;
+    }
 
-        $routes = new RouteCollectionBuilder($loader);
-        $this->configureRoutes($routes);
+    private function getParameterClassName(\ReflectionMethod $method, int $position): ?string
```

Comment:
we'd better not add this method on the trait: private methods on traits are still covered by the BC policy

Reply:
Okay, makes sense.